doc/neps/ufunc-overrides.rst

```diff
@@ -213,83 +370,128 @@ They have indirect interactions, however, because NumPy's
 most numerical classes, the easiest way to override binary operations is
 thus to define ``__array_ufunc__`` and override the corresponding
 Ufunc. The class can then, like :class:`ndarray` itself, define the
-binary operators in terms of Ufuncs. Here, one has to take some care.
-E.g., the simplest implementation would be::
+binary operators in terms of Ufuncs. Here, one has to take some care to
+ensure that one allows for other classes to indicate they are not
+compatible, i.e., implementations should be something like::
 
     class ArrayLike(object):
+        ...
         def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
             ...
             return result
-        ...
-        def __mul__(self, other):
-            return self.__array_ufunc__(np.multiply, '__call__', self, other)
-
-Suppose now, however, that ``other`` is class that does not know how to
-deal with arrays and ufuncs, but does know how to do multiplication::
 
-    class MyObject(object):
-        __array_ufunc__ = None
+        # Option 1: call ufunc directly
```

Comment:
I updated [my gist with the mixin](https://gist.github.com/shoyer/36b84ab064f027df318c0b823558de24) to follow @mhvk's option 2 and @pv's guidance about type hierarchies for `ArrayLike`. I kept a redundant check for `other.__array_ufunc__ is None` in binary operations, because we *really* don't want to people to try to write array types that are above classes that set `__array_ufunc__ = None` in the type-hierarchy, and this precludes the possibility of making that mistake with a poor `__array_ufunc__` implementation.

I agree that we should recommend only copying the behavior that we choose for numpy.ndarray itself, but one additional consideration (especially for the mixin class) is that option 2 works even in versions of NumPy prior to `__array_ufunc__` but option 1 does not. So possibly we would recommend the mixin approach in my gist for people who need more backwards compatibility and implement a different mixin for numpy 1.13+ that can assume that ufunc dispatch to `__array_ufunc__` works.